### PR TITLE
[Inference API] Add Completion Inference API for Alibaba Cloud AI Search Model

### PR DESCRIPTION
**Pull request by Huaixinww**
Related to https://github.com/elastic/elasticsearch/pull/111181
Add Completion Inference API for Alibaba Cloud AI Search Model.
### Prerequisites to Model Creation
An Alibaba Cloud Account with Alibaba Cloud Opensearch access
An api key used to access Alibaba Cloud AI Search Model

### Inference Model Creation:

```
PUT _inference/completion/{inference_model_id}
{
  "service": "alibabacloud-ai-search",
  "service_settings": {
    "api_key": "{{api_key}}",
    "service_id": "<<service_id>>",
    "host": "<<host>",
    "workspace": "<<workspace_name>>",
    <<ADDITIONAL SERVICE SETTINGS (see below)>>,
  }
  "task_settings": {
    <<TASK SETTINGS (see below)>>
  }
}
```

### Testing
Creating the inference endpoint for Alibaba Cloud AI Search

```
PUT _inference/completion/os-completion-test
{
    "service": "alibabacloud-ai-search",
    "service_settings": {
      "host" : "default-j01.platform-cn-shanghai.opensearch.aliyuncs.com",
      "api_key": "{{API_KEY}}",
      "service_id": "ops-qwen-turbo",
      "workspace" : "default"
    }
}
```

Performing completion inference:

```
POST _inference/completion/os-completion-test
{
  "input":["What is Elastic?"]
}
```

### CODE REVIEW

**Comment by weizijun** [comment]
Hi, @davidkyle, we have an LLM called qwen in Alibaba Cloud, and we provide inference services for it, here is the Completion Inference API, can you help review it?
Here is the document: https://help.aliyun.com/zh/open-search/search-platform/developer-reference/text-generation-api-details

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by davidkyle** [comment]
Thank you @weizijun, I'm excited another contribution from you.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                )
+            );
+            return;
+        }
+
+        var docsOnlyInput = (DocumentsOnlyInput) inferenceInputs;
+        if (docsOnlyInput.getInputs().size() % 2 == 0) {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    "Alibaba Completion's inputs must be an odd number, the last input is the current query.",
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java:68]: I don't understand this restriction.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        this.failedToSendRequestErrorMessage = constructFailedToSendRequestMessage(null, "AlibabaCloud Search completion");
+        this.requestCreator = AlibabaCloudSearchCompletionRequestManager.of(account, model, serviceComponents.threadPool());
+    }
+
+    @Override
+    public void execute(InferenceInputs inferenceInputs, TimeValue timeout, ActionListener<InferenceServiceResults> listener) {
+        if (inferenceInputs instanceof DocumentsOnlyInput == false) {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    format("Invalid inference input type, task type [%s] do not support Field [query]", TaskType.RERANK),
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java:57]:
```suggestion
                    format("Invalid inference input type, task type [%s] do not support Field [query]", TaskType.COMPLETION),
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/completion/AlibabaCloudSearchCompletionRequestEntity.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        {
+            builder.startArray(MESSAGE);
+            {
+                for (int i = 0; i < messages.size(); i++) {
+                    builder.startObject();
+                    {
+                        String roleValue = i % 2 == 0 ? ROLE_USER : ROLE_ASSISTANT;
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/alibabacloudsearch/completion/AlibabaCloudSearchCompletionRequestEntity.java:46]: `AlibabaCloudSearchCompletionAction` has a check that there are a odd number of inputs. Is that related to this code here?

**Review by davidkyle — COMMENTED** [review]
The code looks good. I have a question about why the input must be an odd number of strings, I don't understand the reason for that. Thank you.

**Comment by Huaixinww** [comment]
> The code looks good. I have a question about why the input must be an odd number of strings, I don't understand the reason for that. Thank you.

@davidkyle 
This is because Alibaba Cloud's completion API supports the functionality for historical conversations.

For example:
```
curl -XPOST -H"Content-Type: application/json" 
"http://****-hangzhou.opensearch.aliyuncs.com/v3/openapi/workspaces/default/text-generation/ops-qwen-turbo" 
-H "Authorization: Bearer ${API-KEY}"   
 -d "{
      \"messages\":[
      {
          \"role\":\"user\",
          \"content\":\"Where is the capital of Henan?\"    # history query
      },
      {
          \"role\":\"assistant\",
          \"content\":\"Zhengzhou\"                          # history answer
      },
      {
          \"role\":\"user\",
          \"content\":\"What fun things are there?\"       # current query
      }
      ],
      \"stream\":false
}"
```
For details, you can refer to the messages parameter in the body of this document: https://help.aliyun.com/zh/open-search/search-platform/developer-reference/text-generation-api-details

Due to limitations in the inference framework, currently in our implementation, if there are more than one input parameter, we process the first n inputs as history queries and history answers, and use the last parameter as the current query, so we check the number of inputs.

we are also considering placing the historical Q&A in task_settings with a format similar to the one below:
```
{
	"task_settings":{
		"history":[
			{
				"role":"${role}",
				"content":"${content}"
			},
			{
				"role":"${role}",
				"content":"${content}"
			}
			...
		]
	}
}
```
This way, the input parameters will only require the current query to be filled in.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/completion/AlibabaCloudSearchCompletionTaskSettings.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+     * @param requestTaskSettings the settings passed in within the task_settings field of the request
+     * @return a constructed {@link AlibabaCloudSearchCompletionTaskSettings}
+     */
+    public static AlibabaCloudSearchCompletionTaskSettings of(
+        AlibabaCloudSearchCompletionTaskSettings originalSettings,
+        AlibabaCloudSearchCompletionTaskSettings requestTaskSettings
+    ) {
+        return new AlibabaCloudSearchCompletionTaskSettings(
+            requestTaskSettings.getParameters() != null ? requestTaskSettings.getParameters() : originalSettings.getParameters()
+        );
```
[the remark this answers is not in the document]
Comment (Huaixinww) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/alibabacloudsearch/completion/AlibabaCloudSearchCompletionTaskSettings.java]: thanks for ur opinion! I've refactor this part to merge this parameter.

**Comment by Huaixinww** [comment]
hi~ @davidkyle 
If you find the input restrictions confusing, we can limit it to just one input for completion, similar to what Cohere does.

**Comment by davidkyle** [comment]
Thanks @Huaixinww, your idea for adding history is very inventive given the restrictions of the inference API design, this is a very imaginative solution. In future we want to expose all the options in the Alibaba API so the user can explicitly set the message history and any other options. 

Starting with your example, I ran this completion:

```
POST _inference/completion/ali-chat
{
  "input":["Where is the capital of Henan?", "The capital of Henan is Zhengzhou.", "What fun things are there?" ]
}
```

And the history is clearly considered in the response. 
```
{
  "completion": [
    {
      "result": "I'm sorry, I do not have enough information to provide a specific list of fun things to do in Zhengzhou, Henan. I can only tell you that Zhengzhou is the capital of Henan province. To find out about fun activities, attractions, or events in Zhengzhou, I would suggest researching local tourism websites, asking locals, or checking out travel guides for the area."
    }
  ]
}
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+                )
+            );
+            return;
+        }
+
+        var docsOnlyInput = (DocumentsOnlyInput) inferenceInputs;
+        if (docsOnlyInput.getInputs().size() % 2 == 0) {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    "Alibaba Completion's inputs must be an odd number, the last input is the current query.",
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java:68]: The error message could explain that the request and response history can be included in the input.

```suggestion
                    "Alibaba Completion's inputs must be an odd number. The last input is the current query, all preceding inputs are the completion history as pairs of user input and the assistants response.",
```

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by davidkyle** [comment]
@elasticmachine update branch

**Comment by davidkyle** [comment]
@elasticmachine test this please

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            );
+            return;
+        }
+
+        var docsOnlyInput = (DocumentsOnlyInput) inferenceInputs;
+        if (docsOnlyInput.getInputs().size() % 2 == 0) {
+            listener.onFailure(
+                new ElasticsearchStatusException(
+                    "Alibaba Completion's inputs must be an odd number. The last input is the current query, "
+                        + "all preceding inputs are the completion history as pairs of user input and the assistants response.",
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/action/alibabacloudsearch/AlibabaCloudSearchCompletionAction.java:69]:
```suggestion
                        + "all preceding inputs are the completion history as pairs of user input and the assistant's response.",
```

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by Huaixinww** [comment]
@elasticmachine test this please

**Comment by davidkyle** [comment]
@elasticmachine update branch

**Comment by davidkyle** [comment]
@elasticmachine test this please

**Comment by davidkyle** [comment]
@szabosteve please can you update the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/infer-service-alibabacloud-ai-search.html) with the new completion task
